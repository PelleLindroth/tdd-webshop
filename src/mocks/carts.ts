export interface ICart {
  userLogin: string
  items: ICartItem[]
}

export interface ICartItem {
  productId: string // id f√∂r den produkt som refereras till
  amount: number // antal produkter av den sorten
}

export const cartsDb: ICart[] = [
  { userLogin: 'bananpaj', items: [{ productId: '1', amount: 1 }] },
  { userLogin: 'grillkorv', items: [{ productId: '3', amount: 2 }] },
  // {
  //   userLogin: 'paco',
  //   items: [
  //     { productId: '2', amount: 1 },
  //     { productId: '1', amount: 3 },
  //   ],
  // },
]
