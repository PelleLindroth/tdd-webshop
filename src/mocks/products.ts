export interface IProduct {
  id: string // unikt värde för varje produkt
  name: string // produktens namn
  price: number // produktens pris i SEK
}

export const productsDb: IProduct[] = [
  { id: '1', name: 'book', price: 120 },
  { id: '2', name: 'table', price: 1400 },
  { id: '3', name: 'car', price: 150000 },
  { id: '4', name: 'desk', price: 1200 },
  { id: '5', name: 'laptop', price: 7800 },
  { id: '6', name: 'tv', price: 6920 },
]

export const getAllProducts = (): IProduct[] => {
  return productsDb
}

export const getProductById = (id: string): IProduct | undefined => {
  const product = productsDb.find((item) => item.id === id)

  return product
}
