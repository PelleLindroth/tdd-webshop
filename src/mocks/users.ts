export interface IUser {
  email: string
  password: string
}

export const usersDb: IUser[] = [
  {
    email: 'pelle@yahoo.com',
    password: 'grillkorv',
  },
  {
    email: 'emma@yahoo.com',
    password: 'bananpaj',
  },
  {
    email: 'renzo@yahoo.com',
    password: 'paco',
  },
]


export const validateUser = (email: string, password: string) => {
  const user = usersDb.find(user => user.email === email)
  if (user) {
    return user.password === password ? user : null
  } else {
    return null
  }
} 