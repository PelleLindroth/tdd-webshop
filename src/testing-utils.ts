import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'

const renderWithRouter = (ui: ReactElement, options?: any) => render(ui, { wrapper: MemoryRouter, ...options })

export { renderWithRouter }
