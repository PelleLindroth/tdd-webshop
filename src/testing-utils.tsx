import { render } from '@testing-library/react'
import { ReactElement } from 'react'
import { MemoryRouter } from 'react-router-dom'
import { Routes, Route } from 'react-router'

const renderWithRouter = (ui: ReactElement, options?: any) =>
  render(ui, { wrapper: MemoryRouter, ...options })

const withPath = (entry: string, element: ReactElement, path: string) => (
  <MemoryRouter initialEntries={[entry]}>
    <Routes>
      <Route path={path} element={element} />
    </Routes>
  </MemoryRouter>
)
export { renderWithRouter, withPath }
