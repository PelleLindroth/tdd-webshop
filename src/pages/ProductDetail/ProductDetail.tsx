import { IProduct } from '../../mocks/products'
import { useParams } from 'react-router'

const ProductDetail = (props: {
  products: IProduct[]
  cart: IProduct[]
  setCart: React.Dispatch<React.SetStateAction<IProduct[]>>
}) => {
  const { products, cart, setCart } = props

  const { id } = useParams()

  const product = products.find((prod) => prod.id === id)

  const handleAddToCart = () => {
    if (product) {
      setCart([...cart, product])
    }
  }

  return (
    <>
      {product ? (
        <div>
          <h1>{product.name}</h1>
          <h2>SEK{product.price}</h2>
          <button onClick={handleAddToCart}>Add to cart</button>
        </div>
      ) : (
        <h1>No product found</h1>
      )}
    </>
  )
}

export default ProductDetail
