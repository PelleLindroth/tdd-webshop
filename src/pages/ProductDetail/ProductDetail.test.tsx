import { shallow } from 'enzyme'
import ProductDetail from './ProductDetail'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { productsDb } from '../../mocks/products'
import { renderWithPath } from '../../testing-utils'

describe('ProductDetail', () => {
  it('renders ProductDetail component correctly', () => {
    const wrapper = shallow(
      <ProductDetail cart={[]} setCart={jest.fn()} products={productsDb} />
    )

    expect(wrapper).toMatchSnapshot()
  })
  it('renders product detail page for a book when params id is 1', async () => {
    renderWithPath(
      '/product/1',
      <ProductDetail cart={[]} setCart={jest.fn()} products={productsDb} />,
      '/product/:id'
    )

    const title = await screen.findByText(/book/i)
    expect(title).toBeInTheDocument()
  })
  it("renders a not found message if product with matching id doesn't exist", () => {
    renderWithPath(
      '/product/25',
      <ProductDetail cart={[]} setCart={jest.fn()} products={productsDb} />,
      '/product/:id'
    )

    const message = screen.queryByText(/no product/i)

    expect(message).toBeInTheDocument()
  })
  it('renders an Add to Cart button', async () => {
    renderWithPath(
      '/product/2',
      <ProductDetail cart={[]} setCart={jest.fn()} products={productsDb} />,
      '/product/:id'
    )

    const button = await screen.findByRole('button', { name: /add to cart/i })

    expect(button).toBeInTheDocument()
  })
  it('calls setCart when button is clicked', async () => {
    const setCartSpy = jest.fn()

    renderWithPath(
      '/product/2',
      <ProductDetail cart={[]} setCart={setCartSpy} products={productsDb} />,
      '/product/:id'
    )

    const button = await screen.findByRole('button', { name: /add to cart/i })

    userEvent.click(button)

    expect(setCartSpy).toHaveBeenCalled()
    expect(setCartSpy).toHaveBeenCalledTimes(1)
  })
})

// describe("Product detail integration tests", () => {
//   it("updates cart in header component when button is clicked", () => {
//     // render App
//     // click on product
//     // click add to cart
//     // check header component updates
//   })
// })

// product matches params id - renders correct product

// 404 no product found if no product with matching id exists

// Renders an Add to cart button

// Adds products to cart when add button is clicked

// Integration test - header cart updates when button is clicked
