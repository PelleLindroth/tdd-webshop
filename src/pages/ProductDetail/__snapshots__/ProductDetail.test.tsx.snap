// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail renders ProductDetail component correctly 1`] = `
<Fragment>
  <h1>
    No product found
  </h1>
</Fragment>
`;
